### PR TITLE
feat(syncRef): enhance syncRef type restrict

### LINKED ISSUE #3511: syncRef: type constraint not precise
### Describe the bug

When left and right types are incompatible, the expected type constraint is that they must implement options.direction corresponding to options.transform. However, in reality, options are optional, which may lead to unexpected runtime behavior.

``` typescript
const num = ref(0) // Ref<number>
const bool = ref(true) // Ref<boolean>

// Expected(in typescript): compile-time error
// Actual: allowed
syncRef(num, bool) 

num.value++ // Causes "bool" to be assigned a number type
expect(bool).toBeTypeOf('boolean') // Error

bool.value = false // Causes "num" to be assigned a boolean type
expect(num).toBeTypeOf('number') // Error
```

Is this a compromise for some necessary scenario?

##### The most straightforward way:

``` typescript
type SyncRefOptions<L, R = L> = ConfigurableFlushSync & {
  deep?: boolean
  immediate?: boolean
    
} & IfBothExtend<L, R, {
  direction?: 'both' | 'ltr' | 'rtl'
  transform?: Transform<L, R>
}, {
  direction?: 'both'
  transform: Required<Transform<L, R>>
} | {
  direction: 'ltr'
  transform: Transform<L, R> & { ltr: unknown }
} | {
  direction: 'rtl'
  transform: Transform<L, R> & { rtl: unknown }
}>

type IfBothExtend<T, U, A, B = never> = T extends U ? U extends T ? A : B : B
interface Transform<L, R> {
  ltr?: (left: L) => R
  rtl?: (right: R) => L
}


export function syncRef<L, R = L>(left: Ref<L>, right: Ref<R>, options: SyncRefOptions<L, R> ) :() => void
export function syncRef<T>(left: Ref<T>, right: Ref<T>, options?: SyncRefOptions<T> ) :() => void
```

##### Or consider treating direction as a generic:

``` typescript
type SyncRefOptions<L, R = L, D extends Direction = 'both'> = ConfigurableFlushSync & {
  deep?: boolean
  immediate?: boolean
    
  direction?: D
} & {
  transform?: Transform<L, R>
} & IfBothExtend<L, R, {}, {
  transform: Pick<Required<Transform<L, R>>, D extends 'both' ? 'ltr' | 'rtl' : D>
}>

type IfBothExtend<T, U, A, B = never> = T extends U ? U extends T ? A : B : B
type Direction = 'ltr' | 'rtl' | 'both'
interface Transform<L, R> {
  ltr?: (left: L) => R
  rtl?: (right: R) => L
}

export function syncRef<T>(left: Ref<T>, right: Ref<T>, options?: SyncRefOptions<T>): () => void
export function syncRef<L, R = L, D extends Direction = 'both'>(left: Ref<L>, right: Ref<R>, options: SyncRefOptions<L, R, D>): () => void
```

#### After the fix, the behavior aligns with expectations:

``` typescript
const num = ref(0) // Ref<number>
const bool = ref(true) // Ref<boolean>

// @ts-expect-error
syncRef(num, bool)()

// @ts-expect-error
syncRef(num, bool, { transform: {
  rtl: right => Number(right),
} })()

// @ts-expect-error
syncRef(num, bool, { transform: {
  ltr: left => !!left,
} })()

// allowed
syncRef(num, bool, { transform: {
  ltr: left => !!left,
  rtl: right => Number(right),
} })()
```

### Reproduction

https://stackblitz.com/edit/vitejs-vite-t3mvtk/?file=src%2FApp.vue,src%2Ftest.ts

### System Info

```Shell
System:
OS: Linux 5.0 undefined
CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
Memory: 0 Bytes / 0 Bytes
Shell: 1.0 - /bin/jsh
Binaries:
Node: 18.18.0 - /usr/local/bin/node
Yarn: 1.22.19 - /usr/local/bin/yarn
npm: 9.4.2 - /usr/local/bin/npm
pnpm: 8.9.2 - /usr/local/bin/pnpm
npmPackages:
@vueuse/core: ^10.5.0 => 10.5.0 
vue: ^3.2.25 => 3.2.31
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't [already an issue](https://github.com/vueuse/vueuse/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a VueUse issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/core instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vueuse/vueuse/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by Doctor-wu**
fix#3511

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vueuse/vueuse/blob/main/packages/guidelines.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

<details>
<summary><strong>⚠️ Slowing down new functions</strong></summary>
<br>

> **Warning**: **Slowing down new functions**
>
> As the VueUse audience continues to grow, we have been inundated with an overwhelming number of feature requests and pull requests. As a result, maintaining the project has become increasingly challenging and has stretched our capacity to its limits. As such, in the near future, we may need to slow down our acceptance of new features and prioritize the stability and quality of existing functions. **Please note that new features for VueUse may not be accepted at this time.** If you have any new ideas, we suggest that you first incorporate them into your own codebase, iterate on them to suit your needs, and assess their generalizability. If you strongly believe that your ideas are beneficial to the community, you may submit a pull request along with your use cases, and we would be happy to review and discuss them. Thank you for your understanding.

</details>

---

### Description

fix #3511 
From the set theory perspective to restrict the option's type

### Additional context

---

### <samp>🤖 Generated by Copilot at 8495e56</samp>

This pull request improves the `syncRef` function and its TypeScript support. It adds more type checks and constraints for the function options and arguments, and it covers various use cases with a new test file.

### <samp>🤖 Generated by Copilot at 8495e56</samp>

*  Restrict and infer the types of the `syncRef` function options based on the direction and the ref types ([link](https://github.com/vueuse/vueuse/pull/3515/files?diff=unified&w=0#diff-4d06ccb8d3aefeb967955e826b907fb164dcc8e10889f6e48b03792ae45c3d1eL1-R45), [link](https://github.com/vueuse/vueuse/pull/3515/files?diff=unified&w=0#diff-4d06ccb8d3aefeb967955e826b907fb164dcc8e10889f6e48b03792ae45c3d1eL25-R153))
* Simplify the `syncRef` function implementation and handle the optional options parameter ([link](https://github.com/vueuse/vueuse/pull/3515/files?diff=unified&w=0#diff-4d06ccb8d3aefeb967955e826b907fb164dcc8e10889f6e48b03792ae45c3d1eL49-R165))
* Add a test case for the `syncRef` function in `packages/shared/syncRef/index.test.ts` ([link](https://github.com/vueuse/vueuse/pull/3515/files?diff=unified&w=0#diff-00f6d1bad81095b21247b8aeef7ed3fef414bd55869280b7a144a87b5613cc4fR100-R244))

### CODE REVIEW

File: packages/shared/syncRef/index.ts [hunk 1]
```diff
@@ -22,36 +61,108 @@ export interface SyncRefOptions<L, R = L> extends ConfigurableFlushSync {
...
+ * 4. A ∩ B = ∅
  * @param left
  * @param right
+ * @param args
  */
-export function syncRef<L, R = L>(left: Ref<L>, right: Ref<R>, options: SyncRefOptions<L, R> = {}) {
+export function syncRef<L, R, D extends Direction = 'both'>(
+  left: Ref<L>,
+  right: Ref<R>,
+  ...args: Equal<L, R> extends true
```
Comment (qq3063994579) [inline packages/shared/syncRef/index.ts:150]: ...[options = {} as SyncRefOptions<L, R, D>]: Equal<L, R> extends true
    ? [options?: SyncRefOptions<L, R, D>]
    : [options: SyncRefOptions<L, R, D>]

更好的参数提示
better parameter tips

Comment (Doctor-wu) [reply packages/shared/syncRef/index.ts:150]: make sense, I will add a commit to enhance it

**Review by qq3063994579 — COMMENTED** [review]
From machine translation

``` typescript
// It seems that this implementation has better readability, but some details need further consideration.
export type SyncRefOptions<
  L,
  R,
  D extends Direction,
> = ConfigurableFlushSync & {
  /**
   * Watch deeply
   *
   * @default false
   */
  deep?: boolean;
  /**
   * Sync values immediately
   *
   * @default true
   */
  immediate?: boolean;

  /**
   * Direction of syncing. Value will be redefined if you define syncConvertors
   *
   * @default 'both'
   */
  direction?: D;
} & TransformFinal<L, R, D>;

type IfNever<T, A, B> = [T] extends [never] ? A : B;

// Determine the direction with data updates based on 'direction (D)'.
type DirectionNeedSync<D extends Direction> = D extends "both"
  ? "ltr" | "rtl"
  : D;

// Determine the direction with incompatible types when updating data based on 'L' and 'R'.
type DirectionIncompatibleType<L, R> =
  | (L extends R ? never : "ltr")
  | (R extends L ? never : "rtl");

// When there are data updates in a certain direction, and the types of data updates in that direction are incompatible, it requires the implementation of the corresponding transformation method.
type DirectionRequired<L, R, D extends Direction> = Extract<
  DirectionNeedSync<D>,
  DirectionIncompatibleType<L, R>
>;

type TransformFinal<L, R, D extends Direction> = {
  transform?: Partial<Transform<L, R>>;
  // transform?: Partial<Pick<Transform<L, R>, DirectionNeedSync<D>>>; Needs further consideration.
} & IfNever<
  DirectionRequired<L, R, D>,
  {},
  { transform: Pick<Transform<L, R>, DirectionRequired<L, R, D>> }
>;
```


中文

``` typescript
// 似乎这种实现的可读性更好一些，但部分细节有待推敲
export type SyncRefOptions<
  L,
  R,
  D extends Direction,
> = ConfigurableFlushSync & {
  /**
   * Watch deeply
   *
   * @default false
   */
  deep?: boolean;
  /**
   * Sync values immediately
   *
   * @default true
   */
  immediate?: boolean;

  /**
   * Direction of syncing. Value will be redefined if you define syncConvertors
   *
   * @default 'both'
   */
  direction?: D;
} & TransformFinal<L, R, D>;

type IfNever<T, A, B> = [T] extends [never] ? A : B;

// 根据 direction(D)，确定有数据更新的方向
type DirectionNeedSync<D extends Direction> = D extends "both"
  ? "ltr" | "rtl"
  : D;

// 根据 L 和 R，确定数据更新时类型不兼容的方向
type DirectionIncompatibleType<L, R> =
  | (L extends R ? never : "ltr")
  | (R extends L ? never : "rtl");

// 当存在某方向的数据更新，且该方向的数据更新的类型不兼容时，则要求实现对应的转换方法
type DirectionRequired<L, R, D extends Direction> = Extract<
  DirectionNeedSync<D>,
  DirectionIncompatibleType<L, R>
>;

type TransformFinal<L, R, D extends Direction> = {
  transform?: Partial<Transform<L, R>>;
  // transform?: Partial<Pick<Transform<L, R>, DirectionNeedSync<D>>>; 有待推敲
} & IfNever<
  DirectionRequired<L, R, D>,
  {},
  { transform: Pick<Transform<L, R>, DirectionRequired<L, R, D>> }
>;
```

**Comment by Doctor-wu** [comment]
Such clean codes! But in your way, it can't restrict users from passing more directions such as 
```typescript
    syncRef(ref0, ref1, {
      direction: 'rtl',
      transform: {
        ltr: v => v, // useless here, maybe it's users wrong, but if we can hint users, it will be great
      },
    })()
```

I will try to use a more elegant way to implement 😂

**Comment by Doctor-wu** [comment]
> Such clean codes! But in your way, it can't restrict users from passing more directions such as
> 
> ```ts
>     syncRef(ref0, ref1, {
>       direction: 'rtl',
>       transform: {
>         ltr: v => v, // useless here, maybe it's users wrong, but if we can hint users, it will be great
>       },
>     })()
> ```
> 
> I will try to use a more elegant way to implement 😂

And now we have a cleaner implement 🎉

**Comment by Doctor-wu** [comment]
> > Such clean codes! But in your way, it can't restrict users from passing more directions such as
> > ```ts
> >     syncRef(ref0, ref1, {
> >       direction: 'rtl',
> >       transform: {
> >         ltr: v => v, // useless here, maybe it's users wrong, but if we can hint users, it will be great
> >       },
> >     })()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I will try to use a more elegant way to implement 😂
> 
> From machine translation
> 
> I made changes to my previous code, and this time it passed the test
> 
> ```ts
> type Equal<A, B> = A extends B ? (B extends A ? true : false) : false
> 
> type IfNever<T, A, B> = [T] extends [never] ? A : B
> 
> export type SyncRefOptions<
>   L,
>   R,
>   D extends Direction,
> > = ConfigurableFlushSync & {
>   /**
>    * Watch deeply
>    *
>    * @default false
>    */
>   deep?: boolean
> 
>   /**
>    * Sync values immediately
>    *
>    * @default true
>    */
>   immediate?: boolean
> 
>   /**
>    * Direction of syncing. Value will be redefined if you define syncConvertors
>    *
>    * @default 'both'
>    */
>   direction?: D
> } & TransformFinal<L, R, D>
> 
> type Direction = 'ltr' | 'rtl' | 'both'
> 
> interface Transform<L, R> {
>   ltr: (left: L) => R
>   rtl: (right: R) => L
> }
> 
> // Determine the direction of data updates based on the 'direction' (D).
> type DirectionNeedSync<D extends Direction> = D extends 'both' ? 'ltr' | 'rtl' : D
> 
> // Determine the direction of data updates that are incompatible based on the types of L and R.
> type DirectionIncompatibleType<L, R> =
>   | (L extends R ? never : 'ltr')
>   | (R extends L ? never : 'rtl')
> 
> // When there are data updates in a certain direction, and the types of data updates in that direction are incompatible,
> // it requires the implementation of the corresponding transformation methods.
> type DirectionRequired<L, R, D extends Direction> = Extract<
>   DirectionNeedSync<D>,
>   DirectionIncompatibleType<L, R>
> >
> 
> // If D doesn't extend any type, TypeScript will consider that the 'transform' property doesn't exist properties
> // (even though you can check the existence of the property using 'in', TypeScript would infer the property as 'unknown').
> type TransformFinal<L, R, D extends Direction> = D extends any ? ({
>   transform?: Pick<Partial<Transform<L, R>>, DirectionNeedSync<D>>
> } & IfNever<
>   DirectionRequired<L, R, D>,
>   { },
>   { transform: Pick<Transform<L, R>, DirectionRequired<L, R, D>> }
> >) : never
> ```

That's a suitable implement, but I prefer the set theory, I think it's more scientific 😂

**Comment by qq3063994579** [comment]
> > > Such clean codes! But in your way, it can't restrict users from passing more directions such as
> > > ```ts
> > >     syncRef(ref0, ref1, {
> > >       direction: 'rtl',
> > >       transform: {
> > >         ltr: v => v, // useless here, maybe it's users wrong, but if we can hint users, it will be great
> > >       },
> > >     })()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I will try to use a more elegant way to implement 😂
> > 
> > 
> > From machine translation
> > I made changes to my previous code, and this time it passed the test
> > ```ts
> > type Equal<A, B> = A extends B ? (B extends A ? true : false) : false
> > 
> > type IfNever<T, A, B> = [T] extends [never] ? A : B
> > 
> > export type SyncRefOptions<
> >   L,
> >   R,
> >   D extends Direction,
> > > = ConfigurableFlushSync & {
> >   /**
> >    * Watch deeply
> >    *
> >    * @default false
> >    */
> >   deep?: boolean
> > 
> >   /**
> >    * Sync values immediately
> >    *
> >    * @default true
> >    */
> >   immediate?: boolean
> > 
> >   /**
> >    * Direction of syncing. Value will be redefined if you define syncConvertors
> >    *
> >    * @default 'both'
> >    */
> >   direction?: D
> > } & TransformFinal<L, R, D>
> > 
> > type Direction = 'ltr' | 'rtl' | 'both'
> > 
> > interface Transform<L, R> {
> >   ltr: (left: L) => R
> >   rtl: (right: R) => L
> > }
> > 
> > // Determine the direction of data updates based on the 'direction' (D).
> > type DirectionNeedSync<D extends Direction> = D extends 'both' ? 'ltr' | 'rtl' : D
> > 
> > // Determine the direction of data updates that are incompatible based on the types of L and R.
> > type DirectionIncompatibleType<L, R> =
> >   | (L extends R ? never : 'ltr')
> >   | (R extends L ? never : 'rtl')
> > 
> > // When there are data updates in a certain direction, and the types of data updates in that direction are incompatible,
> > // it requires the implementation of the corresponding transformation methods.
> > type DirectionRequired<L, R, D extends Direction> = Extract<
> >   DirectionNeedSync<D>,
> >   DirectionIncompatibleType<L, R>
> > >
> > 
> > // If D doesn't extend any type, TypeScript will consider that the 'transform' property doesn't exist properties
> > // (even though you can check the existence of the property using 'in', TypeScript would infer the property as 'unknown').
> > type TransformFinal<L, R, D extends Direction> = D extends any ? ({
> >   transform?: Pick<Partial<Transform<L, R>>, DirectionNeedSync<D>>
> > } & IfNever<
> >   DirectionRequired<L, R, D>,
> >   { },
> >   { transform: Pick<Transform<L, R>, DirectionRequired<L, R, D>> }
> > >) : never
> > ```
> 
> That's a suitable implement, but I prefer the set theory, I think it's more scientific 😂

Yes, your implementation looks more scientifically rigorous. But it's too hard for me to read😂, perhaps because I don't understand the type well enough.

**Comment by Doctor-wu** [comment]
It's okay, as long as it works normally, why don't you contact me on Twitter, maybe we can make friends

**Review by antfu — APPROVED** [review]

**Pull request merged**